### PR TITLE
Add episode 14

### PR DESCRIPTION
**Pull request by bitsondatadev**

### CODE REVIEW

File: _episodes/14.md [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+</p>
+
+## Iceberg links
+
+ * [Apache Iceberg](https://iceberg.apache.org/)
+ * [Apache Iceberg Community](https://iceberg.apache.org/community/)
+
+## Guests
+  
+ * David Phillips, creator of Trino and Presto, and CTO at
```
Comment (mosabua) [inline _episodes/14.md:47]: "creator of Trino/Presto" .. otherwise this looks like they are different tools..

File: _episodes/14.md [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+data was to, in some way shape or form, use the Hive model. The Hive model is
+very simple, but it enabled running queries over files in a distributed file
+system. 
+
+To accomplish this, Hive uses a metastore service which stores and manages
+metadata. For Hive and Trino, this metadata acts as a pointer to the files
+containing the data, contains the file format, and has the column structure and
+types. This enabled Hive to query the correct files and data within those files
+for a SQL query. For more information on Hive's architecture, read the
+[Gentle Introduction to Hive](/blog/2020/10/20/intro-to-hive-connector.html)
```
Comment (mosabua) [inline _episodes/14.md:86]: use the post_url syntax instead.

File: _episodes/14.md [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+blog. After the initial model gained adoption, Hive added other features such as
+partitioning, which used the directory structures of the filesystems to split
```
Comment (mosabua) [inline _episodes/14.md:88]: partitioning. It uses ...

File: _episodes/14.md [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+the files of data partitioned on a special column into different directories.
+We talk about this in more depth [a few episodes back](/episodes/5.html).
+
+This solved some initial issues facing engineers in big data, but there were
```
Comment (mosabua) [inline _episodes/14.md:92]: The Hive model solved

File: _episodes/14.md [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+quite a few issues with this model. The Hive model is very rigid and not able
```
Comment (mosabua) [inline _episodes/14.md:93]: It is very rigid..

File: _episodes/14.md [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+to adapt to your needs as requirements change. For example, if you started
+partitioning your data splitting by date and segmenting by month, that table is
+stuck with that partitioning forever. The only way to update it is to create a
+new table with your new partition values, and migrate all of your data from
+the old table to the new table. Another issue stems from the separation of data
```
Comment (mosabua) [inline _episodes/14.md:98]: to the new table. With the common data sizes such a migration is often a long process, sometimes even impossible. Another issue...

File: _episodes/14.md [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+stored in the metastore and data stored in the file system. The source of many
+issues in Hive relate to the Hive metastore getting out of sync. A third but
```
Comment (mosabua) [inline _episodes/14.md:100]: in Hive is caused by the Hive metastore..

File: _episodes/14.md [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+new table with your new partition values, and migrate all of your data from
+the old table to the new table. Another issue stems from the separation of data
+stored in the metastore and data stored in the file system. The source of many
+issues in Hive relate to the Hive metastore getting out of sync. A third but 
+not final issue, is that running operations against the metastore is a timely
+process when running operations like list files on more modern object storage. 
+
+As all these problems amassed over the years, it was clear something needed
+to be done. In the last few years, a few candidate table formats have come to
+the forefront of data engineering trends. Apache Iceberg, Apache Hudi, and the
```
Comment (mosabua) [inline _episodes/14.md:106]: trends. Examples are Apache Iceberg...

File: _episodes/14.md [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+proprietary Databricks' Deltalake. The goal of these systems was to modernize
+the old Hive data structure. To Trino, Iceberg was particularly promising due to
+the list of promising features like schema versioning support and hidden
+partitioning that made it particularly attractive. However, the biggest element
+that sets Iceberg apart is the specification they are writing around it. This
```
Comment (mosabua) [inline _episodes/14.md:111]: is the written specification for it.

File: _episodes/14.md [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+### The Iceberg table format
+
+Iceberg, is a new table format developed at Netflix that aims to replace older 
+table formats like Hive to add better flexibility as the schema evolves, atomic
+operations, speed, and just dependability. To be clear, it's not a new file
+format, as it still uses ORC, Parquet, and Avro, but a table format. Netflix 
+donated Iceberg to the Apache Software Foundation and it is now a top level
+project! 
+
+Iceberg handles both the data on disk just like Hive, but instead it stores the
```
Comment (mosabua) [inline _episodes/14.md:126]: Somewhere in this paragraph mention that these are AVRO files..

File: _episodes/14.md [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+Hive is that the manifest files keep track of table data at the file level as
+opposed to directory level that Hive uses. By doing so, Iceberg avoids having to
+list all files in a directory, which becomes a very common and expensive
+operation.
+ 
+<p align="center">
+  <img align="center" width="60%" height="100%" src="/assets/episode/14/iceberg-metadata.png"/><br/>
+</p>
+ 
+By tracking files this way, we not only get begger performance from object
```
Comment (mosabua) [inline _episodes/14.md:141]: better performance

File: _episodes/14.md [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+storage, it also enables serializeable isolation. This addresses the lack of
+consistency between the metadata and file state experienced in Hive.
+
+One of the greater advantages to Iceberg over Hive is the in-place table
+evolution. This means that you can add, drop, or rename a column, as well as, 
+reorder and update a column without any expensive refactoring of tables or 
+moving data around and there will be no adverse effects on your data or
+metadata.
+
+Partition evolution and hidden partitions are particularly invaluable. Before
```
Comment (mosabua) [inline _episodes/14.md:151]: Remove the sentence "Before in Hive, you had.." since this was already discussed earlier.

File: _episodes/14.md [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+in Hive, you had a very strict way you had to model consistent partitioning
+across an entire table. In Iceberg, the *partiion spec* is a description of how 
+to partition data in a table consisting of a list of source columns and
+transforms. Once the spec is created, it generates a partition tuple that is
+applied uniformly to the files created with that spec. Unlike Hive, that
+requires you to modify and send a special column that acts as the partiion
```
Comment (mosabua) [inline _episodes/14.md:157]: partition

File: _episodes/14.md [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+spec generated in the Java API.
+
+```
+PartitionSpec spec = PartitionSpec.builderFor(schema)
+        .hour("event_time")
+        .identity("level")
+        .build();
+```
+
+In the example above, we create a separate hourly partition on the `event_time`
```
Comment (mosabua) [inline _episodes/14.md:168]: The example creates a separate hourly...

File: _episodes/14.md [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+
+<p align="center">
+  <img align="center" width="60%" height="100%" src="/assets/episode/14/partition-spec-evolution.png"/><br/>
+</p>
+
+If all that wasn't enough, you can also do time travel and version rollback with
+Iceberg. As we mentioned above, Iceberg keeps track of various snapshots of 
+your data in time through manifest files. As long as you keep those older 
+snapshots around, the files associated with those snapshots will stick around
+as well. This allows you to move around to previous views of the data which can
```
Comment (mosabua) [inline _episodes/14.md:185]: of the data. This is useful for testing, recovery, and many other purposes.

File: _episodes/14.md [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+Iceberg is also able to offer fast scan planning by filtering out the metadata
+files that are irrelevant to the scan, and using the partition spec to only find
+files containing responses to the data. Iceberg filters the metadata using
+partition value ranges and seeing if that is contained within the files of
+the metadata. Then while processing the list of manifest files, Iceberg will
+filter files by query predicates included in the partition, then apply column
+stats to help prune out files that don't match. Iceberg also uses multiple
+concurrent writers to speed things up as a final measure.
+
+Saving the best for last; Iceberg is a community standard and has
```
Comment (mosabua) [inline _episodes/14.md:199]: This repeats the spec content from earlier. Remove one of them or consolidate into one.

File: _episodes/14.md [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+stats to help prune out files that don't match. Iceberg also uses multiple
+concurrent writers to speed things up as a final measure.
+
+Saving the best for last; Iceberg is a community standard and has 
+[a full specification](https://iceberg.apache.org/spec/) which is a nice change
+from Hive which is an ad-hoc specification that people adhere to in some ways.
+There have been many issues over the years due to the variance of how the
+unwritten specification gets interpreted.
+
+### Future development for the Trino Iceberg connector
```
Comment (mosabua) [inline _episodes/14.md:205]: This should probably go after the PR that created the connector.

File: _episodes/14.md [hunk 18]
```diff
@@ -0,0 +1,330 @@
...
+[PR of the week](https://github.com/trinodb/trino/pull/1067).
+
+[Release 318](/docs/current/release/release-318.html), 
+introduced this code that supported querying tables from Apache Iceberg in
+Trino. While the code existed, the Iceberg connector code wasn't officially
+released or documented until a little over a year later in 
+[release 341](/docs/current/release/release-341.html) once the connector reached
+maturity. 
+
+## Demo: Creating tables through Iceberg and reading them through Trino
```
Comment (mosabua) [inline _episodes/14.md:228]: tables with Iceberg and reading the data in Trino

File: _episodes/14.md [hunk 19]
```diff
@@ -0,0 +1,330 @@
...
+
+For this weeks' demo, I wanted to play around with the Iceberg Java API directly.
+You also have the option to use Trino, Spark, or other to ingest and query the
+data, but I wanted to use vanilla Iceberg API's to experience the API and
+hopefully solidify my learning of Iceberg concepts in the process. Make sure you
+follow the instructions in the repository if you don't have Docker or Java
+installed.
+
+Let's start up a local Trino coordinator and hive metastore. Clone this
```
Comment (mosabua) [inline _episodes/14.md:237]: Hive

File: _episodes/14.md [hunk 20]
```diff
@@ -0,0 +1,330 @@
...
+you can check to see that the table in Trino exists in the metastore under
+TABLE_PARAMS. But, if run `SHOW TABLES IN iceberg.logging;` you'll notice that
+the table doesn't show up due to [the issue we discussed above](https://github.com/trinodb/trino/issues/1592).
+
+Let's update the TABLE_PARAMS entry in the metastore db and then query the table
+again.
+
+## Question of the week: Why do we still depend on the Hive metastore if metadata for Iceberg saves to the filesystem?
+
+We kept the metastore as many tests run around using the metastore that exist
```
Comment (mosabua) [inline _episodes/14.md:262]: Who is "we" in this context .. I think this should be changed to either 

The Iceberg community

or 

The Trino community 

or whatever is correct.

File: _episodes/14.md [hunk 21]
```diff
@@ -0,0 +1,330 @@
...
+make the metastore the initial method of use in Iceberg that got developed as
+most developers initially would be migrating from their existing Hive catalog
+and we wanted this transition to use existing tested components.
+
+Currently, the metastore isn't used the same way as in Hive. Trino stores a
+top-level directory that points to the metadata manifest file location and other
+statistics around the table in the TABLE_PARAMS table of the metastore. There
+is a [pull request created by Jack Ye](https://github.com/trinodb/trino/pull/6977)
+to migrate away from using the metastore, but in the meantime, we must still
+use a separate metastore.
```
Comment (mosabua) [inline _episodes/14.md:274]: separate Hive metastore.

also .. in general with Iceberg.. 

or just when using Iceberg with Trino?

File: _episodes/14.md [hunk 22]
```diff
@@ -0,0 +1,330 @@
...
+
+## Events, news, and various links
```
Comment (mosabua) [inline _episodes/14.md:276]: Above this add section about Iceberg tipping ... its mostly written up in the Gdoc

**Review by mosabua — CHANGES_REQUESTED** [review]
A bunch of things to be changed. Generally I think it might be good to cut down on the usage of past tense and future tense a bit more.

File: _episodes/14.md (hunk 11 above)
Comment (bitsondatadev) [reply _episodes/14.md:141]: LOL

File: _episodes/14.md (hunk 20 above)
Comment (bitsondatadev) [reply _episodes/14.md:262]: maybe just Trino as a system

Comment (mosabua) [reply _episodes/14.md:266]: I am not sure what is true..

File: _episodes/14.md [hunk 23]
```diff
@@ -0,0 +1,346 @@
...
+ 4. Now draw the iceberg in the image above that Commander Bun Bun is on.
+ 
+When drawing the iceberg like the image with Commander Bun Bun, the iceberg tips
+over. Careful Commander Bun Bun! It looks like the Apache logo wins! Shout out 
+to [Joshua Tauberer](https://twitter.com/JoshData) for the web page. Shout out 
+to [Megan Thompson-Munson](https://twitter.com/GlacialMeg) for the tweet that 
+started the page. Shout out to 
+[Barton Wright](https://www.linkedin.com/in/bartonwright/) from Manfred's team 
+of writers for being the geek to find this. Shout out to 
+[Ali](https://twitter.com/aliLoney) for being and sport and setting Command Bun
```
Comment (mosabua) [inline _episodes/14.md:289]: being a good sport and

**Review by mosabua — APPROVED** [review]
One more minor tweak but essentially good to go now.

**Pull request merged**